### ISSUE TITLE
Improve Squiz.WhiteSpace.SemicolonSpacing

### ISSUE THREAD

**Issue by vladyslavstartsev** [squizlabs/PHP_CodeSniffer#2426]
Right now it works  mostly correct, but you can see the problem when you start calling chaining methods, for example
where it detects error
```php
$qb
    ->select()
    ->from()
    ->where()
;
```
fixed to 
```php
$qb
    ->select()
    ->from()
    ->where();
```
the other example is
```php
(new Foo())->addBar((new Bar())
    ->addBarData('Data 1')
    ->addBarData('Data 2')
);
```
to 
```php
(new Foo())->addBar((new Bar())
    ->addBarData('Data 1')
    ->addBarData('Data 2'));
```

so the 2 examples where the error was found is false-positive (for me)

Also to be clear, this code should work as it works now (I mean if there is chained method, but all method calls on one line)
```$chainedObject->foo()->bar()->baz();```
I think the above example is perfectly fine, but when we transform it to each method call on each line, it should look like this
```php
$chainedObject
    ->foo()
    ->bar()
    ->baz()
;
```

**Comment by jrfnl** [comment]
@vladyslavstartsev I can reproduce the behaviour with the first example, but the second example would not be touched by this sniff.

The sniff expects **no space** before the semi-colon independently of the code before it (with the exception of the first semicolon within a `for` condition as it may have to comply to other rules).

I suppose a `ignoreNewlines` property could be added, similar to other sniffs, but that wouldn't be limited to chained calls, so the following code would then also be left alone:
```php
$a = $b

;
```

**Comment by vladyslavstartsev** [comment]
@jrfnl yes, I've tried 2nd example yet another time and saw that it works fine if I do,
 not this
```php
(new Foo())->addBar((new Bar())
    ->addBarData('Data 1')
    ->addBarData('Data 2')
);
```
but this
```php
(new Foo())->addBar(
    (new Bar())
        ->addBarData('Data 1')
        ->addBarData('Data 2')
);
```

------

With `ignoreNewlines` it's easy to go, but what I think it's would be better to do something line `ignoreChainedMethods` and then detect if the previous code was chained method, but that's a little bit harder to implement. So with `ignoreChainedMethods` it will work perfectly for the example you've shown

**Comment by jrfnl** [comment]
@vladyslavstartsev The problem you then get is that one person may want it to be ignored for chained methods, another for anonymous classes and yet someone else for multi-line strings.

What you are looking for sounds more like something for which you'd need to write a custom sniff or extend the existing sniff and do some pre-processing and bow out before calling the `parent::process()` method if a chained method is detected.
